Multiple fixes to link the confirmed trip timeline correctly + update composite trips properly

shankari opened the pull request:
This is a pretty significant rewrite of the confirmed places code
And adds some new updates to the composite trips code.

Lots of UI testing done, automated tests forthcoming
@JGreenlee @sebastianbarry for your information and review

shankari committed:
Create and link the timeline properly, similar to the `CLEAN_AND_RESAMPLE` stage

Before this, the confirmed trip and place creation was somewhat ad-hoc:
- we first had only confirmed trips, no timeline, so all the links were broken
- we then added confirmed place, but didn't make sure that links were all set properly
- we completely ignored confirmed untracked
- we switched to using a timeline to get alternating places and trips, but then
  we didn't update the last ending place when a new trip came in

This rewrites the code to follow a similar pattern to the `CLEAN_AND_RESAMPLE`
stage, which we know works.

This is a fix for https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1500543312

We find the last cleaned place and create or update the correspoding confirmed place.
We then iterate over the cleaned trips, update the start place, create a new
end place and link all three of them better.

We also included confirmed_untracked so that the full timeline is created
properly, which involved supporting the confirmed_untracked in multiple code
locations

Also switched the entry creation from new Entry to make_entry so that the
metadata is filled in correctly, similar to 418cffccbb58e2ac990c1d5a677ea2156c06abd9

Testing done:
- Before this change, when running the pipeline twice, the last place of the
  first run is not updated (https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1500543312)
- After this change, when running the pipeline twice, the last place of the first run is updated
    (https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502750747)

shankari committed:
Fix the backwards compat to generate correct confirmed object timelines

In 1c3f128717c6b8374585eacc9dbf417bda5235b9 we added a hack to add confirmed
places to previously created confirmed trips so that the composite trip
creation would work.

However, this hack used a separate field (confirmed_place) and did not properly
link the confirmed trip into the full place/trip timeline.

We fix the hack to correctly create confirmed objects and map from the cleaned
objects to the confirmed objects and use the confirmed untracked properly.

This fixes: https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502015267
Testing done: https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502750747

shankari committed:
Update composite trips when confirmed objects are updated

Before the composite trip changes, we retrieved confirmed trips for display.
So if/when the confirmed trips were updated, we would see them on the next phone retrieval.
In other words, we displayed the terminal output of the pipeline.

However, with the composite trip code, the terminal output is the composite trip.
But we update the confirmed trips as part of the matching.
So if we get new user inputs or update the confirmed place exit timestamps,
they will not be reflected in the display. We are no longer displaying the
terminal output of the pipeline, we have another display layer on top.

So when we update the confirmed objects, we need to figure out which the
matching composite objects are and update them as well.

We generally don't include forward references in our data objects, only backward references.
We already have a reference from the composite trip to the associated end_confirmed_place
We also add a reference to the associated confirmed_trip

Then, when we update a confirmed object, we find the related composite object
and update it as well.

This fixes: https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1500798479
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1500799602
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1503756634
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1504109750

Testing done:

- Loaded data, no additions
- Added place additions via UI
- Pushed changes
- Re-ran pipeline

Additions match

```
>>> pd.json_normalize(all_places)["data.additions"]
0                                                   []
1                                                   []
2    [{'_id': 64361902f5622167bcbdf190, 'user_id': ...
3    [{'_id': 64361902f5622167bcbdf1c2, 'user_id': ...
4    [{'_id': 64361902f5622167bcbdf208, 'user_id': ...
5    [{'_id': 64361902f5622167bcbdf246, 'user_id': ...
6                                                   []
7                                                   []
8                                                   []
9                                                   []
```

Also in the composite trips

```
>>> pd.json_normalize(all_ct)["data.end_confirmed_place.data.additions"]
0                                                   []
1    [{'_id': 64361902f5622167bcbdf190, 'user_id': ...
2    [{'_id': 64361902f5622167bcbdf1c2, 'user_id': ...
3    [{'_id': 64361902f5622167bcbdf208, 'user_id': ...
4    [{'_id': 64361902f5622167bcbdf246, 'user_id': ...
5                                                   []
6                                                   []
7                                                   []
8                                                   []
```

Added trip inputs and additions via UI
Pushed changes
Re-ran pipeline; both show up in the confirmed trips

```
>>> pd.json_normalize(all_ct)["data.additions"]
0                                                   []
1                                                   []
2                                                   []
3                                                   []
4    [{'_id': 64361c6bf5622167bcbe08cf, 'user_id': ...
5    [{'_id': 64361c6bf5622167bcbe0921, 'user_id': ...
6                                                   []
7                                                   []
8                                                   []

>>> for ct in all_ct:
...     print(ct["data"]["user_input"])
...
{}
{}
{}
{}
{'trip_user_input':...}
{'trip_user_input':...}
{'trip_user_input':...}
{}
{}
```

shankari committed:
Add unit tests to check the new functions in composite trip creation

First set of tests for `updateConfirmedAndComposite`
The tests are coded as sub-tests of one giant test so that we don't have to
load data and run the pipeline multiple times.

```
testSetConfirmedTripUserInput DONE
.
----------------------------------------------------------------------
Ran 1 test in 32.686s

OK
```

sebastianbarry commented on emission/analysis/plotting/composite_trip_creation.py:
```diff
@@ -10,43 +11,88 @@
+    # that was removed as inconsistent with the data model
+    # https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502015267
+    import emission.core.get_database as edb
+    curr_confirmed_place_count = edb.get_analysis_timeseries_db().count_documents({"metadata.key": "analysis/confirmed_place"})
```
For consistency with the other keys, we may want to have the key `"analysis/confirmed_place"` come from `esda` and add a new key in `analysis_timeseries_queries.py` that is something along the lines of:
```python
CONFIRMED_PLACE_KEY = "analysis/confirmed_place"
```

shankari replied:
there is already a `CONFIRMED_PLACE_KEY`
This came in because I copied-pasted from an interactive session.
Good catch, will fix.

sebastianbarry commented on emission/analysis/plotting/composite_trip_creation.py:
```diff
@@ -10,43 +11,88 @@
+    # https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502015267
+    import emission.core.get_database as edb
+    curr_confirmed_place_count = edb.get_analysis_timeseries_db().count_documents({"metadata.key": "analysis/confirmed_place"})
+    if "additions" in ct["data"] and "trip_addition" not in ct["data"]:
```
is this how we are determining whether an entry is "old style"? Why do old style entries have `trip_addition` rather than `addition`?

shankari replied:
https://github.com/shankari/e-mission-server/commit/9eb812d6e48a6abf15824be4e05722ef8c72c7d6

On staging, we will reset the entire pipeline.
But we deployed a previous staging version to production for a couple of projects and I know empirically that they are from the version that created `trip_addition`

sebastianbarry commented on emission/analysis/userinput/matcher.py:
```diff
@@ -85,56 +85,107 @@ def create_confirmed_objects(user_id):
+    for ctrip in timeline.trips:
+        # create the trip-like object
+        logging.debug("trip-like entry = %s" % ctrip)
+        if ctrip['metadata']['key'] == esda.EXPECTED_TRIP_KEY:
+            curr_confirmed_trip = create_confirmed_entry(ts, ctrip, esda.CONFIRMED_TRIP_KEY, keys)
+        elif ctrip['metadata']['key'] == esda.CLEANED_UNTRACKED_KEY:
+            curr_confirmed_trip = create_confirmed_entry(ts, ctrip, esda.CONFIRMED_UNTRACKED_KEY, keys)
```
This part got me a little stuck during reading. I believe I understand what is going on, but a comment explaining why we are checking explicitly for either ctrips that contain EXPECTED_TRIP_KEY or CLEANED_UNTRACKED_KEY could help for code readability
```python
# This is where we convert the trip entries in the list to confirmed trips, and if the entry is untracked then we processed the untracked trip without populating all the necessary  confirmed_object_data
```

shankari replied:
This depends on an understanding of the data model.
The trip-like inputs to this timeline are EXPECTED TRIP and CLEANED_UNTRACKED
If you can clarify what you don't understand here, I can add a new comment.

sebastianbarry reviewed (commented):
Looks very good! I only had a few minor comments and mainly questions but besides that the structure looks great. I also though how you approached composite_trips was very clever, I think it is much smarter to have composite_trips contain id links to its respective place object id's and trip object id's as opposed to saving the trip data within the composite_trip object (this is better for modularity and clarity since we are storing id's all around instead of only for certain object types)

sebastianbarry commented:
@shankari Would you like me to test it as well, or have you done your own testing for this code?

shankari commented:
@sebastianbarry thank you for the review.

Unfortunately this is not actually true:

> I think it is much smarter to have composite_trips contain id links to its respective place object id's and trip object id's as opposed to saving the trip data within the composite_trip object 

The composite trip contains the full trip data *in addition to* the ids. Think: why did we choose to use composite trips in the first place? Would storing IDs achieve those goals?

sebastianbarry commented:
> The composite trip contains the full trip data _in addition to_ the ids. Think: why did we choose to use composite trips in the first place? Would storing IDs achieve those goals?

Thank you for clarifying! I see, there is the entire section where we are using `convert_cleaned_to_confirmed` specifically so that we can contain confirmed trip/place data within the composite_trips timeline.

We chose to use composite trips with a few design decisions in mind, namely:

1. so that the trip objects can have their associated place objects contained together. We do this for a few reasons:
    * we only need to make one call to the server to give us "composite_trips", instead of having to do 2 separate calls "get me all the confirmed trips for this time window" and "get me all the confirmed places for this time window". This does not do much in terms of overhead, since the slowdown is caused in the retrieval from the database, see below for reasoning
    * less complexity on the phone code, we do not need to link the places and trips on the phone and it happens in the server code
2. so that the trip geojson location points are automatically returned in a truncated form on the phone. this is so that we dont receive an ID (in the form of timestamps) from the server on the phone, and then have to do another server call to get the location data points and truncate them

We found that the largest time spent in retrieving trip objects was in retrieving the data for the server from the database. We chose purposefully to contain the raw data within the composite_trip object because not containing the data and only containing the IDs would defeat the purpose of storing IDs as we would need to wait on the database data return either way.

shankari committed:
Expand the `save_ground_truth` script to also support saving a composite trip list

The old diary function is retained until we remove the diary code
We use subcommands since we only need the date for the diary and not for the composite trip list
(for which we just download all trips)

Testing done:
- Generated composite trip lists for unit tests

shankari committed:
Fix population of start place additions when grafting timeline

- we were filling in `confirmed_start_place["additions"]` but it should
  actually be in `data`
+ Also improve a log message to help debug this in the future

shankari committed:
Check in a new unit test to test incremental updates of the composite trips

- Load data for the 4th
- Run pipeline
- Load inputs for the 5th
- Run pipeline (they all match to the last place)
- Load data for the 5th
- Run pipeline (they all spread out)
- Load inputs (including trip inputs) for the 4th
- Run pipeline (they all match)

This essentially reproduces
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1502750747

Screenshots of the various test stages:
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506229672

shankari committed:
Multiple fixes to the hack code

- we should get the confirmed place and composite trip count before starting to create composite trips
    otherwise the counts will be steadily increasing
- we should only create the start place or the end place for each trip, not both
    - creating both will lead to duplicate trips
    - we handle this by checking whether the confirmed place corresponding to a cleaned place already
        exists and fixing it
- fixed some start/end mismatches caused by copy pasting

Testing done:
- Reset to an older version of the code
- Ran the pipeline
- Moved ahead to this code
- Re-ran the pipeline
- verified that the object counts were correct
- verified that the resulting composite trips matched the expected values from fb6a98261b9f64f1a1c953c7384c315ec3169584

https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506261786
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506379248
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506398843
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506410568
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506420678
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506431222

shankari commented:
Automated tests done and tested:
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506229672

Hack to add confirmed places tested manually:
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506398843
https://github.com/e-mission/e-mission-docs/issues/880#issuecomment-1506431222

shankari committed:
Change deprecated `remove` to `delete_many`

Untested fallout of
https://github.com/e-mission/e-mission-docs/issues/856
